clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -578,41 +587,45 @@ private boolean shouldStopDrainBatchesForPartition(ProducerBatch first, TopicPar
-                        // the transaction manager track the batch so as to ensure that sequence ordering is maintained
-                        // even if we receive out of order responses.
-                        batch.setProducerState(producerIdAndEpoch, transactionManager.sequenceNumber(batch.topicPartition), isTransactional);
-                        transactionManager.incrementSequenceNumber(batch.topicPartition, batch.recordCount);
-                        log.debug("Assigned producerId {} and producerEpoch {} to batch with base sequence " +
-                                "{} being sent to partition {}", producerIdAndEpoch.producerId,
-                            producerIdAndEpoch.epoch, batch.baseSequence(), tp);
-
-                        transactionManager.addInFlightBatch(batch);
-                    }
-                    batch.close();
-                    size += batch.records().sizeInBytes();
-                    ready.add(batch);
+                // do the rest of the work by processing outside the lock
+                // close() is particularly expensive
+                batch = deque.pollFirst();
+            }
 
-                    batch.drained(now);
-                }
+            boolean isTransactional = transactionManager != null && transactionManager.isTransactional();
+            ProducerIdAndEpoch producerIdAndEpoch =
+                transactionManager != null ? transactionManager.producerIdAndEpoch() : null;
+            if (producerIdAndEpoch != null && !batch.hasSequence()) {
+                // If the producer id/epoch of the partition do not match the latest one
+                // of the producer, we update it and reset the sequence. This should be
+                // only done when all its in-flight batches have completed. This is guarantee
+                // in `shouldStopDrainBatchesForPartition`.
+                transactionManager.maybeUpdateProducerIdAndEpoch(batch.topicPartition);
+
+                // If the batch already has an assigned sequence, then we should not change the producer id and
+                // sequence number, since this may introduce duplicates. In particular, the previous attempt
+                // may actually have been accepted, and if we change the producer id and sequence here, this
+                // attempt will also be accepted, causing a duplicate.
+                //
+                // Additionally, we update the next sequence number bound for the partition, and also have
+                // the transaction manager track the batch so as to ensure that sequence ordering is maintained
+                // even if we receive out of order responses.
+                batch.setProducerState(producerIdAndEpoch, transactionManager.sequenceNumber(batch.topicPartition), isTransactional);
+                transactionManager.incrementSequenceNumber(batch.topicPartition, batch.recordCount);
```

Comment:
I think the expectation is that these 2 would be atomic (i.e. would be bad if one thread executed 615, then another thread executed 615 again and got the same sequence number, before the first thread got a chance to execute 616).
Also I think the expectation is that batches that are ordered one after another in the queue would get the sequence numbers in the same order (i.e. that batch that is later in the queue would get higher sequence number).
Previously these expectations were protected by the queue lock so "poll", "get sequence", "update sequence" would execute as atomic block, with this change the operations could interleave.

Reply:
OK, this makes sense. I agree it makes sense to push this back into the synchronized block; the more important part for reducing lock duration is to move the `close()` outside of the block. Would you agree with this approach?

Reply:
I've addressed this in 34008bf.

Below is flamegraph that `getDeque()` and `close()` consume the CPU in the `drain()` tree. I believe with both of these performed outside of the lock we are still in a good spot here.

![image](https://user-images.githubusercontent.com/3196528/152459130-b14031b4-2f99-463b-940b-8de9b2fb6135.png)

Reply:
Moving `close` outside of locked scope LGTM

Reply:
Can you please test with Java 11 or newer? Looks like you tested with Java 8 which used the slower crc32c method.

Reply:
Correct, this application is jdk8. I'll have to find a jdk11 app to compare, it could take a while as this app is not ready for it. Is a jdk11 test a blocker?

Reply:
No, not a blocker.